### PR TITLE
beacon-chain/execution: no need to reread and unmarshal the eth1Data twice

### PR DESCRIPTION
**Pull request by jsvisa**
**What type of PR is this?**

Other

**What does this PR do? Why is it needed?**

The powChain data maybe large(100MB on goerlia testnet, maybe larger on mainnet), and it consumed much memory to unmarshal the protobuf bytes into `ETH1ChainData`, and 
 inside the `ensureValidPowchainData`, we will also read powchain date, so no need to read(unmarshal) the same data twice.

### CODE REVIEW

File: beacon-chain/execution/service.go [hunk 1]
```diff
@@ -787,23 +783,23 @@ func validateDepositContainers(ctrs []*ethpb.DepositContainer) bool {
 
 // Validates the current powchain data is saved and makes sure that any
 // embedded genesis state is correctly accounted for.
-func (s *Service) ensureValidPowchainData(ctx context.Context) error {
```
Comment (rkapka) [inline beacon-chain/execution/service.go:790]: Because we are returning something now, I would change the name of the function. `validPowchainData` maybe?

Comment (jsvisa) [reply beacon-chain/execution/service.go:790]: good idea

File: beacon-chain/execution/service.go [hunk 2]
```diff
@@ -819,9 +815,9 @@ func (s *Service) ensureValidPowchainData(ctx context.Context) error {
 			Trie:              s.depositTrie.ToProto(),
 			DepositContainers: s.cfg.depositCache.AllDepositContainers(ctx),
 		}
-		return s.cfg.beaconDB.SaveExecutionChainData(ctx, eth1Data)
+		return nil, s.cfg.beaconDB.SaveExecutionChainData(ctx, eth1Data)
```
Comment (nisdas) [inline beacon-chain/execution/service.go:818]: This becomes an issue as now you have returned a nil object, so in the event of new nodes, they will be initialized with a nil eth1data object and presumably panic.

Comment (jsvisa) [reply beacon-chain/execution/service.go]: good catch, fixed.

**Review by nisdas — DISMISSED** [review]
Thanks for this !

**Review by rkapka — APPROVED** [review]

**Comment by rkapka** [comment]
Hey @jsvisa , `TestService_ReceiveBlock` seems to be constantly failing

**Comment by jsvisa** [comment]
@rkapka thanks for the message, but it seems I can't reproduce that locally as below:


```bash
$ go test -tags develop ./beacon-chain/blockchain/ -run TestService_ReceiveBlock  -v
# github.com/prysmaticlabs/prysm/v4/beacon-chain/blockchain.test
/bin/ld: warning: static_code.o: missing .note.GNU-stack section implies executable stack
/bin/ld: NOTE: This behaviour is deprecated and will be removed in a future version of the linker
=== RUN   TestService_ReceiveBlock
=== RUN   TestService_ReceiveBlock/applies_block_with_state_transition
=== RUN   TestService_ReceiveBlock/saves_attestations_to_pool
=== RUN   TestService_ReceiveBlock/updates_exit_pool
=== RUN   TestService_ReceiveBlock/notifies_block_processed_on_state_feed
--- PASS: TestService_ReceiveBlock (1.26s)
    --- PASS: TestService_ReceiveBlock/applies_block_with_state_transition (0.18s)
    --- PASS: TestService_ReceiveBlock/saves_attestations_to_pool (0.19s)
    --- PASS: TestService_ReceiveBlock/updates_exit_pool (0.16s)
    --- PASS: TestService_ReceiveBlock/notifies_block_processed_on_state_feed (0.17s)
=== RUN   TestService_ReceiveBlockUpdateHead
--- PASS: TestService_ReceiveBlockUpdateHead (0.42s)
=== RUN   TestService_ReceiveBlockBatch
=== RUN   TestService_ReceiveBlockBatch/applies_block_with_state_transition
=== RUN   TestService_ReceiveBlockBatch/notifies_block_processed_on_state_feed
--- PASS: TestService_ReceiveBlockBatch (0.49s)
    --- PASS: TestService_ReceiveBlockBatch/applies_block_with_state_transition (0.10s)
    --- PASS: TestService_ReceiveBlockBatch/notifies_block_processed_on_state_feed (0.10s)
PASS
ok      github.com/prysmaticlabs/prysm/v4/beacon-chain/blockchain       2.274s
```

Do I need to use `bazel` to run the test?

**Comment by nisdas** [comment]
Once https://github.com/prysmaticlabs/prysm/pull/12953 is merged in, it should unblock this PR from being merged in.

**Comment by nisdas** [comment]
@jsvisa It appears that tests constantly fail for the PR, if its possible can you test the failing package with bazel ?

```
bazel test //beacon-chain/blockchain:go_raceoff_test
```

**Comment by jsvisa** [comment]
> @jsvisa It appears that tests constantly fail for the PR, if its possible can you test the failing package with bazel ?
> 
> ```
> bazel test //beacon-chain/blockchain:go_raceoff_test
> ```

thanks for the information, I'll try to run it on my PC, and let you know if I can find why

**Comment by jsvisa** [comment]
@nisdas please take another look, thanks.

**Review by nisdas — APPROVED** [review]
Thanks for fixing this !

**Pull request merged**